### ISSUE TITLE
Deserialize fallback on error

### ISSUE THREAD

**Issue by pashadia** [jonasbb/serde_with#276]
Is there a way to specify a fallback type to deserialize from? Or perhaps even better, a list of types (all implementing DeserializeAs) that could be tried in order when deserializing?

Specifically, my use case involves parsing some JSON values which might be either numbers, or numbers encoded as strings. Either `{"value": 123 }` or `{"value": "123"}` are possible.

Of course, writing a specific type for this case is trivial.

But I'm thinking about an interface like:
```
struct Container {
    #[serde_as(as = "DeserializeOptions<Self, DisplayFromStr>")]
    value: usize,
}

fn main() {
    let foo_json = r#"{ "value": 123 }"#;
    let foo: Container = serde_json::from_str(foo_json).unwrap();
    assert_eq!(foo.value, 123);

    let bar_json = r#"{ "value": "123" }"#;
    let bar: Container = serde_json::from_str(bar_json).unwrap();
    assert_eq!(bar.value, 123);
}
```

This also would simplify the implementation of various structs like `BytesOrString`

I'm definitely willing to work on this, if there's no ready available alternative.

**Comment by jonasbb** [comment]
Hi, that is an interesting idea. The `DeserializeOptions` is supposed to be a special case of an untagged serde enum, similar to this, which always returns a `usize` (in this case).

```rust
#[serde_as]
#[derive(serde::Deserialize)]
#[serde(untagged)]
enum Helper {
	A(#[serde_as(as = "_")] usize),
	B(#[serde_as(as = "DisplayFromStr")] usize),
}
```

> Or perhaps even better, a list of types (all implementing DeserializeAs) that could be tried in order when deserializing?

I like the idea of generalizing it to multiple types instead of just 2. For the general case a type list would be necessary. That seems quite complex.
Maybe it is enough to limit it to 4 or maybe 6 choices at most. In that case it might be possible to do this with tuples, i.e., something like `DeserializeOptions<(_,)>`, `DeserializeOptions<(_, DisplayFromStr)>`, `DeserializeOptions<(_, DisplayFromStr, NumberFromBool)>`.

Another question is what is the serialization behavior? Maybe using the first type in the list as the serialization behavior is good enough there.

> I'm definitely willing to work on this,

I'm happy to have contributions here. I'm happy to help out too if you have any questions. Probably best to just create a draft PR if you want to work on it.

---

In your case of deserializing from either a number or string it might be worthwhile to have a specialized type. Simply because this is a common variant.

**Comment by jonasbb** [comment]
The type is available in 1.8.0 as `PickFirst`.
